File: pkg/app/services/trading/order_book_service.ex
defmodule OrderBookService do
  @spec get_volume_at_price(Instrument.name(), TradeOrder.Side.t(), Decimal.t()) :: Decimal.t()
  def get_volume_at_price(instument_name, side, price) do
    [instument_name, side |> Atom.to_string(), price]
    |> DB.query_val("""
      SELECT volume
      FROM price_level
      WHERE side = $2
        AND instrument_id = (SELECT id FROM instrument WHERE name = $1)
        AND price =  $3
    """)
    |> case do
      nil -> 0.0
      val -> val
    end
  end

  @spec get_volumes(Instrument.name(), TradeOrder.Side.t()) :: [
          {PriceLevel.price(), PriceLevel.volume()}
        ]
  def get_volumes(instrument_name, side) do
    [instrument_name, side |> Atom.to_string()]
    |> DB.query_list("""
      SELECT (price, volume)
      FROM price_level
      WHERE side = $2
      AND price > 0
      AND instrument_id = (SELECT id FROM instrument WHERE name = $1)
      ORDER BY price #{case side do
      :SELL -> "ASC"
      :BUY -> "DESC"
    end}
    """)
    |> Enum.map(fn {a, b} -> {Decimal.to_float(a), Decimal.to_float(b)} end)
  end
end

File: pkg/services/order_book/order_book_service.go
package orderbook

import (
	"open-outcry/pkg/db"
	"open-outcry/pkg/models/instrument"
	order "open-outcry/pkg/models/order_book"
	tradeorder "open-outcry/pkg/models/trade_order"
)

func GetVolumeAtPrice(instrumentName instrument.InstrumentName, side tradeorder.OrderSide, price tradeorder.OrderPrice) float64 {
	res := db.QueryVal[float64](`
	 SELECT volume
	 FROM price_level
	 WHERE side = $2
	   AND instrument_id = (SELECT id FROM instrument WHERE name = $1)
	   AND price =  $3
   `, instrumentName, side, price)
	return res
}

func GetVolumes(instrumentName instrument.InstrumentName, side tradeorder.OrderSide) []order.PriceVolume {
	var orderBy string
	switch side {
	case tradeorder.Sell:
		orderBy = "ASC"
	case tradeorder.Buy:
		orderBy = "DESC"
	}
	res := db.QueryList[order.PriceVolume](`
		SELECT price, volume
		FROM price_level
		WHERE side = $2
		AND price > 0
		AND instrument_id = (SELECT id FROM instrument WHERE name = $1)
		ORDER BY price `+orderBy,
		instrumentName,
		side,
	)
	return res
}

func GetOrderBook(instrumentName instrument.InstrumentName) order.OrderBook {
	res := db.QueryList[order.PriceVolume](`
		SELECT price, volume, side
		FROM price_level
		WHERE price > 0
		AND instrument_id = (SELECT id FROM instrument WHERE name = $1)
		ORDER BY price ASC, side DESC
	`, instrumentName)

	book := order.OrderBook{
		SellSide: make([]order.PriceVolume, 0),
		BuySide:  make([]order.PriceVolume, 0),
	}
	for _, entry := range res {
		switch entry.Side {
		case tradeorder.Sell:
			book.SellSide = append(book.SellSide, order.PriceVolume{Price: entry.Price, Volume: entry.Volume})
		case tradeorder.Buy:
			book.BuySide = append(book.BuySide, order.PriceVolume{Price: entry.Price, Volume: entry.Volume})
		}
	}

	for i, j := 0, len(book.BuySide)-1; i < j; i, j = i+1, j-1 {
		book.BuySide[i], book.BuySide[j] = book.BuySide[j], book.BuySide[i]
	}

	return book
}
